torch/csrc/distributed/rpc/request_callback_no_python.cpp

```diff
@@ -583,25 +502,48 @@ void RequestCallbackNoPython::processRpc(
             wrappedMsgType,
             messageId,
             wrappedRpcResponseFuture);
-      }
-      wrappedRpcResponseFuture->addCallback([wrappedRpcResponseFuture,
+
+        auto tid = std::this_thread::get_id();
+        wrappedRpcResponseFuture->addCallback(
+            at::wrapPropagateTLSState<void>([wrappedRpcResponseFuture,
                                              responseFuture,
-                                             profiledEvents =
-                                                 std::move(profiledEvents),
-                                             profilingKeyId] {
-        if (wrappedRpcResponseFuture->hasError()) {
-          // Propagate error
-          responseFuture->setError(wrappedRpcResponseFuture->error()->what());
-        } else {
-          auto rpcWithProfilingResp = std::make_unique<RpcWithProfilingResp>(
-              MessageType::RUN_WITH_PROFILING_RESP,
-              std::move(*wrappedRpcResponseFuture).moveValue(),
-              profiledEvents,
-              profilingKeyId);
-          responseFuture->markCompleted(
-              std::move(*rpcWithProfilingResp).toMessage());
-        }
-      });
+                                             profilingKeyId,
+                                             profilingConfig,
+                                             tid] {
+              std::vector<torch::autograd::profiler::Event> profiledEvents;
+              // Defer consolidation of profiler events until async work has
+              // completed (such as async UDF)
+
+              TORCH_INTERNAL_ASSERT(
+                  torch::autograd::profiler::profilerEnabled(),
+                  "Expected profiler to be enabled!");
+
+              // Only clean up TLS states of profiler if we are disabling on
+              // the main thread.
+              bool shouldCleanUpTLSStates = (std::this_thread::get_id() == tid);
+              auto event_lists = torch::autograd::profiler::disableProfiler(
+                  shouldCleanUpTLSStates, true);
```

Comment:
It looks like we're doing some cleanup as part of the destructor of `torch::autograd::profiler::TLSProfilerGuard g` and then there is some additional cleanup here. Could you explain what is the difference between the two?

Reply:
The destructor in the `TLSProfilerGuard` cleans up the profiler thread local states in the main thread, but does not call `consolidate()` which clears out the event lists, In the continuation thread, we don't clean up thread local states (those are restored by the destructor of `ThreadLocalStateGuard` in `at::wrapPropagateTLSState` but we do call `consolidate()` to get all of the events, even the async ones (which would not have been logged if we called `consolidate()` on main thread).